File: cfd/F-MPI/cfdio.f90
module cfdio

  use mpi

  implicit none

contains

subroutine writedatafiles(psi, m, n, scale, comm)

  integer :: m, n, scale, comm
  double precision ::  psi(0:m+1, 0:n+1)

  double precision, allocatable :: vel(:,:,:)
  integer, allocatable :: rgb(:,:,:)

  double precision :: modvsq, hue
  integer :: size, rank, irank, i, j, k,  ierr, ix, iy

  integer, parameter :: tag = 1, iounitvel = 10, iounitcol = 11

  integer, dimension(MPI_STATUS_SIZE) :: status

  call mpi_comm_size(comm, size, ierr)
  call mpi_comm_rank(comm, rank, ierr)

! Compute local velocities and colours

  allocate(rgb(3,m,n))
  allocate(vel(2,m,n))

  do i = 1, m
     do j = 1, n

        vel(1,i,j) =   (psi(i,j+1)-psi(i,j-1)) / 2.0
        vel(2,i,j) = - (psi(i+1,j)-psi(i-1,j)) / 2.0

        modvsq = vel(1,i,j)**2 + vel(2,i,j)**2
        hue = modvsq**0.4

        call hue2rgb(hue, rgb(1,i,j), rgb(2,i,j), rgb(3,i,j))

     end do
  end do

!  Receive data

  if (rank == 0) then

     open(unit=iounitcol, file='colourmap.dat', form='formatted')
     open(unit=iounitvel, file='velocity.dat',  form='formatted')

     do irank = 0, size-1

        if (irank /= 0) then

           call mpi_recv(rgb, 3*m*n, MPI_INTEGER, irank, tag, &
                         comm, status, ierr)

           call mpi_recv(vel, 2*m*n, MPI_DOUBLE_PRECISION, irank, tag, &
                         comm, status, ierr)

        end if

!  Write out

        do j = 1, n

           iy = irank*n + j

           do i = 1, m

              ix = i

!  Write colour map of velocity magnitude at every point

              write(iounitcol,fmt='(i4,1x,i4,1x,i3,1x,i3,1x,i3)') &
                    ix, iy, rgb(1,i,j), rgb(2,i,j), rgb(3,i,j)

!  Only write velocity vectors every "scale" points

              if (mod(ix-1,scale) == (scale-1)/2 .and. &
                  mod(iy-1,scale) == (scale-1)/2         ) then

                 write(iounitvel,fmt='(i4,1x,i4,1x,g12.5,1x,g12.5)') &
                       ix, iy, vel(1,i,j), vel(2,i,j)
              end if
             
           end do
        end do
     end do

     close(unit=iounitcol)
     close(unit=iounitvel)

  else

     call mpi_ssend(rgb, 3*m*n, MPI_INTEGER, 0, tag, comm, ierr)
     call mpi_ssend(vel, 2*m*n, MPI_DOUBLE_PRECISION, 0, tag, comm, ierr)

  end if

end subroutine writedatafiles


subroutine writeplotfile(m, n, scale)

  integer :: m, n, scale
  integer, parameter :: iounit = 10

  open(unit=iounit, file='cfd.plt', form='formatted')

  write(iounit,*) 'set size square'
  write(iounit,*) 'set key off'
  write(iounit,*) 'unset xtics'
  write(iounit,*) 'unset ytics'

  write(iounit,fmt='('' set xrange ['',i4,'':'',i4, '']'')') 1-scale, m+scale
  write(iounit,fmt='('' set yrange ['',i4,'':'',i4, '']'')') 1-scale, n+scale

  write(iounit,fmt='('' plot "colourmap.dat" w rgbimage, "velocity.dat" u 1:2:&
       &('',i2,''*0.75*$3/sqrt($3**2+$4**2)):&
       &('',i2,''*0.75*$4/sqrt($3**2+$4**2)) &
       &with vectors  lc rgb "#7F7F7F"'')') scale, scale

  close(unit=iounit)

end subroutine writeplotfile


subroutine hue2rgb(hue, r, g, b)

  double precision :: hue

  integer :: r, g, b
  integer, parameter :: rgbmax = 255

  r = rgbmax*colfunc(hue-1.0)
  g = rgbmax*colfunc(hue-0.5)
  b = rgbmax*colfunc(hue    )

end subroutine hue2rgb


double precision function colfunc(x)

  double precision :: x, absx, val

  double precision, parameter :: x1 = 0.2, x2 = 0.5

  absx = abs(x)

  if (absx .gt. x2) then
     val = 0.0
  else if (absx .lt. x1) then
     val = 1.0
  else
     val = 1.0 - ((absx-x1)/(x2-x1))**2
  end if

  colfunc = val
      
end function colfunc

end module cfdio

File: C-MPI/cfdio.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "cfdio.h"
#include "arraymalloc.h"

#ifndef USEVLA
void writedatafiles(int m, int n, double **psi, int scale, MPI_Comm comm)
#else
void writedatafiles(int m, int n, double psi[m+2][n+2], int scale, MPI_Comm comm)
#endif
{
  typedef double vecvel[2];
  typedef int    vecrgb[3];

  vecvel **vel;
  vecrgb **rgb;

  double modvsq, hue;
  int size, rank, irank, i,j, ix, iy;
  int nvel, nrgb;

  int tag=1;

  MPI_Status status;

  MPI_Comm_rank(comm,&rank);
  MPI_Comm_size(comm,&size);

  if (rank==0) printf("\n\nWriting data files ...\n");

  vel = (vecvel **) arraymalloc2d(m,n,sizeof(vecvel));
  rgb = (vecrgb **) arraymalloc2d(m,n,sizeof(vecrgb));

  //calculate velocities and hues

  double v1, v2;

  for (i=0;i<m;i++)
    {
      for (j=0;j<n;j++)
	{
	  vel[i][j][0] =  (psi[i+1][j+2]-psi[i+1][j])/2.0;
	  vel[i][j][1] = -(psi[i+2][j+1]-psi[i][j+1])/2.0;

	  v1 = vel[i][j][0];
	  v2=  vel[i][j][1];

	  modvsq = v1*v1 + v2*v2;

	  hue = pow(modvsq,0.4);

	  hue2rgb(hue,&(rgb[i][j][0]),&(rgb[i][j][1]),&(rgb[i][j][2]));
	}
    }

  //receive ddata

  if (rank == 0)
    {
      FILE *cfile, *vfile;

      cfile=fopen("colourmap.dat","w");
      vfile=fopen("velocity.dat","w");

      for (irank=0;irank<size;irank++)
	{
	  if (irank != 0)
	    {
	      MPI_Recv(&rgb[0][0][0],3*m*n,MPI_INT,irank,tag,comm,&status);
	      MPI_Recv(&vel[0][0][0],2*m*n,MPI_DOUBLE,irank,tag,comm,&status);
	    }

	  for (i=0;i<m;i++)
	      {
		ix = irank*m+i+1;

                for (j=0;j<n;j++)
		  {
		    iy = j+1;

                    fprintf(cfile,"%i %i %i %i %i\n", ix, iy,
			    rgb[i][j][0], rgb[i][j][1],rgb[i][j][2]);

                    if ((ix-1)%scale == (scale-1)/2 &&
			(iy-1)%scale == (scale-1)/2    )
		      {

                        fprintf(vfile,"%i %i %f %f\n",
				ix,iy,vel[i][j][0],vel[i][j][1]);
		      }
		  }
	      }
	  }

        fclose(vfile);
        fclose(cfile);
      }
    else
      {
	MPI_Ssend(&rgb[0][0][0],3*m*n,MPI_INT,0,tag,comm);
	MPI_Ssend(&vel[0][0][0],2*m*n,MPI_DOUBLE,0,tag,comm);
      }

  free(rgb);
  free(vel);

  if (rank ==0) printf("... done!\n");
}

void writeplotfile(int m, int n, int scale)
{
  FILE *gnuplot;

  gnuplot = fopen("cfd.plt","w");

  fprintf(gnuplot,"set size square\n");
  fprintf(gnuplot,"set key off\n");
  fprintf(gnuplot,"unset xtics\n");
  fprintf(gnuplot,"unset ytics\n");

  fprintf(gnuplot,"set xrange [%i:%i]\n",1-scale,m+scale);
  fprintf(gnuplot,"set yrange [%i:%i]\n",1-scale,n+scale);

  fprintf(gnuplot,"plot \"colourmap.dat\" w rgbimage, \"velocity.dat\" u 1:2:(%d*0.75*$3/sqrt($3**2+$4**2)):(%d*0.75*$4/sqrt($3**2+$4**2)) with vectors  lc rgb \"#7F7F7F\"",scale,scale);

  fclose(gnuplot);

  printf("\nWritten gnuplot script 'cfd.plt'\n");
}


void hue2rgb(double hue, int *r, int *g, int *b)
{
  int rgbmax = 255;

  *r = (int)(rgbmax*colfunc(hue-1.0));
  *g = (int)(rgbmax*colfunc(hue-0.5));
  *b = (int)(rgbmax*colfunc(hue    ));
}


double colfunc(double x)
{
  double absx;

  double x1=0.2;
  double x2=0.5;

  absx=fabs(x);

  if (absx > x2)
    {
      return 0.0;
    }
  else if (absx < x1)
    {
      return 1.0;
    }
  else
    {
      return 1.0-pow((absx-x1)/(x2-x1),2);
    }
}
